internal/rbd/snap_diff.go

```diff
@@ -0,0 +1,195 @@
+/*
+Copyright 2025 The Ceph-CSI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rbd
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	rbderrors "github.com/ceph/ceph-csi/internal/rbd/errors"
+	"github.com/ceph/ceph-csi/internal/rbd/types"
+	"github.com/ceph/ceph-csi/internal/util/log"
+
+	librbd "github.com/ceph/go-ceph/rbd"
+	"github.com/container-storage-interface/spec/lib/go/csi"
+	"google.golang.org/grpc/codes"
+)
+
+// StreamMetadata streams the block metadata for a snapshot.
+// If baseRBDSnap is provided, it calculates the delta between the
+// current snapshot and the base snapshot.
+func (rbdSnap *rbdSnapshot) StreamMetadata(
+	ctx context.Context,
+	sendResponse types.SMSSendResponse,
+	startingOffset int64,
+	maxResults int32,
+	baseSnap types.Snapshot,
+) error {
+	var fromSnapID uint64
+	image, err := rbdSnap.open()
+	if err != nil {
+		return fmt.Errorf("failed to open image %q: %w", rbdSnap, err)
+	}
+	defer func() {
+		cErr := image.Close()
+		if cErr != nil {
+			log.WarningLog(ctx, "resource leak, failed to close image: %v", cErr)
+		}
+	}()
+
+	snapID, err := rbdSnap.getRBDSnapID(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get snapshot ID of %q: %w", rbdSnap, err)
+	}
+
+	err = image.SetSnapByID(snapID)
+	if err != nil {
+		return fmt.Errorf("failed to set snapshot %q by ID %d: %w",
+			rbdSnap, snapID, err)
+	}
+
+	if baseSnap != nil {
+		baseRBDSnap, err := rbdSnapFromSnapshot(baseSnap)
+		if err != nil {
+			return fmt.Errorf("failed to convert base snapshot to rbdSnapshot: %w", err)
+		}
+		fromSnapID, err = baseRBDSnap.getRBDSnapID(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to get ID of base snapshot ID: %w", err)
+		}
+	}
+
+	changedBlocks := make([]*csi.BlockMetadata, 0)
+	cb := createDiffIterateByIDCB(
+		ctx,
+		sendResponse,
+		changedBlocks,
+		maxResults,
+	)
+	diffIterateByIDConfig := librbd.DiffIterateByIDConfig{
+		Offset:   uint64(startingOffset),
+		Length:   uint64(rbdSnap.VolSize),
+		Callback: cb,
+	}
+	if baseSnap != nil {
+		// If a base snapshot is provided, set the FromSnapID to
+		// the base snapshot ID to get the delta between the two snapshots.
+		diffIterateByIDConfig.FromSnapID = fromSnapID
+	}
+
+	diffIterateErr := image.DiffIterateByID(diffIterateByIDConfig)
+	err = handleDiffIterateError(ctx, diffIterateErr)
+	if err != nil {
+		log.ErrorLog(ctx, "failed to get diff: %s", err)
```

Comment:
log it at the caller? we dont need to log this at multiple places in this function?